File: src/languages/indonesian/indonesian.ts
/**
 * Native Indonesian (id) text phonemizer — canonical IPA. Bahasa Indonesia has a shallow,
 * near-phonemic Latin orthography, so this is a rule-based transliterator: digraphs (ng→ŋ, ny→ɲ, sy→ʃ, kh→x)
 * then single letters, ⟨e⟩→schwa [ə] by default, closed-syllable lax allophones (i→ɪ, u→ʊ, e→ɛ, o→ɔ), falling
 * diphthongs ai/au/oi, and syllable-final ⟨k⟩ → glottal stop [ʔ]. Penultimate stress (skips a schwa nucleus).
 */
import type { Phonemizer } from "../../registry.ts";
import type { CountForms } from "../../core/normalizeSymbols.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { loadManifest } from "../../core/loadManifest.ts";
import { loadTsvMap } from "../../core/loadTsv.ts";
import { makeSymbolNormalizer } from "../../core/normalizeSymbols.ts";
import { makeUnreadableTest } from "../../core/initialisms.ts";
import { normalizeIndonesian } from "./normalize.ts";

interface NumbersDef {
    units: string[];
    belas: string;
    puluh: string;
    ratus: string;
    ribu: string;
    juta: string;
    seprefix: string;
    /** The word for the decimal comma ("koma"); absent before, so a decimal ran its two halves together. */
    decimalWord?: string;
}
interface IndonesianDef {
    digraphs: Record<string, string>;
    consonants: Record<string, string>;
    vowels: Record<string, string>;
    laxVowels: Record<string, string>;
    diphthongs: Record<string, string>;
    letterNames: Record<string, string>;
    clausePunctuation: Record<string, string>;
    numbers: NumbersDef;
    phonotactics: { vowels: string; onsets: string[]; codas: string[]; digraphs: string[] };
    /** The shared symbol tier's data — moved verbatim, comments included. See the jsonc. */
    symbolTier: {
        percent: CountForms;
        currency: Record<string, CountForms>;
        units: Record<string, CountForms>;
        magnitudes: string[];
        ampersand: string;
        multiply: { times: string; by?: string };
        exponentWords: { squared: CountForms; cubed: CountForms; position?: "before" | "after" };
    };
}
const DEF = loadManifest<IndonesianDef>(import.meta.url, "indonesian.jsonc");

/** The Indonesian data tables, exported under the name every other language uses for its manifest.
 *  ⚠ Indonesian has no separate manifest.ts — the shape is declared here and loaded here; this alias is so
 *  callers (and tests) do not have to know that. */
export { DEF as MANIFEST };
const CLAUSE_MARK = DEF.clausePunctuation;
const NUM = DEF.numbers;

export type ForeignPhonemizer = (latin: string) => string;

const isVowelLetter = (c: string): boolean => "aeiou".includes(c);
const VOWEL_PH = "aeiouəɛɔɪʊ";

interface Seg {
    ph: string;
    vowelLetter: string; // the source vowel letter (for lax resolution); "" for consonants
}

/** Scan a lowercased Indonesian word into segments (consonants + tense-default vowels). */
function scan(w: string): Seg[] {
    const s = [...w];
    const segs: Seg[] = [];
    for (let i = 0; i < s.length; ) {
        // Falling diphthong at word end (ai/au/oi): the two vowels form one nucleus.
        const di = s[i]! + (s[i + 1] ?? "");
        if (DEF.diphthongs[di] && i + 2 === s.length) {
            segs.push({ ph: DEF.diphthongs[di]!, vowelLetter: "" });
            i += 2;
            continue;
        }
        const dg = s[i]! + (s[i + 1] ?? "");
        if (DEF.digraphs[dg]) {
            segs.push({ ph: DEF.digraphs[dg]!, vowelLetter: "" });
            i += 2;
            continue;
        }
        const c = s[i]!;
        if (isVowelLetter(c)) {
            segs.push({ ph: DEF.vowels[c]!, vowelLetter: c });
            i++;
        } else if (DEF.consonants[c]) {
            segs.push({ ph: DEF.consonants[c]!, vowelLetter: "" });
            i++;
        } else {
            i++; // unknown → skip
        }
    }
    return segs;
}

// NOTE: Indonesian /i u e o/ have lax closed-syllable allophones [ɪ ʊ ɛ ɔ], but the referee (and standard
// descriptions) apply them only erratically/dialectally — kecil→kətʃil, air→air keep TENSE vowels. We emit the
// tense phonemes for canonical consistency; the eval folds the tense/lax axis (config.ts). laxVowels stays in
// the manifest for reference.

/** Penultimate stress on the second-to-last vowel; if that nucleus is a schwa, shift to the final. */
function stressIndex(segs: Seg[]): number {
    const nuclei = segs
        .map((s, i) => (VOWEL_PH.includes(s.ph[0] ?? "") ? i : -1))
        .filter((i) => i >= 0);
    if (nuclei.length === 0) return -1;
    if (nuclei.length === 1) return nuclei[0]!;
    const penult = nuclei[nuclei.length - 2]!;
    if (segs[penult]!.ph === "ə") return nuclei[nuclei.length - 1]!; // schwa can't bear stress
    return penult;
}

/**
 * Can this letter run be an Indonesian word at all? Indonesian syllables are simple — (C)V(C) with a
 * short coda inventory — so a capital run with no vowel, or with a cluster the language does not
 * license, is an acronym and nothing else.
 */
const isUnreadableIndonesian = makeUnreadableTest({
    vowels: new RegExp(`[${DEF.phonotactics.vowels}]`, "u"),
    legalOnsets: new Set(DEF.phonotactics.onsets),
    legalCodas: new Set(DEF.phonotactics.codas),
    digraphs: new Set(DEF.phonotactics.digraphs),
});

/**
 * ⚠ READABLE AND STILL SPELLED OUT — the cases the phonotactic test gets wrong in the other direction.
 * `core/initialisms.ts` states the principle: "readability is not convention". `AS` (Amerika Serikat)
 * and `REM` are perfectly pronounceable Indonesian shapes and are still said letter by letter, so a
 * test on shape alone would read them as words. Evidence: the FLEURS id_id audio has `as` as
 * `a e s` and `aol` as `a o e l` — spelled, in both cases.
 */
const SPELLED_ANYWAY = new Set(["as", "rem", "oha", "aol"]);

/**
 * ⚠ ACRONYM LEXICON — the word-pronounced borrowings whose reading is NOT derivable from Indonesian
 * orthography, and so has to be listed rather than computed.
 *
 * The readability gate above decides the right QUESTION (word, not letters), but the answer then comes
 * from the native rules — and native ⟨c⟩ is /t͡ʃ/, so `UNESCO` read as an Indonesian word gives
 * *unəst͡ʃˈo*. These borrowings keep the /k/ of their source spelling. That is a fact about each word,
 * not about its shape, which is what a lexicon is for — the same reasoning as the Nguni loan list.
 *
 * ATTESTED IN THE FLEURS id_id AUDIO, not assumed:
 *     unesco   ASR `o n i a o n i s k o`   -> …nesko, with /k/
 *     covid    ASR `k ɔ f i t`             -> /kovid/, with /k/
 *     acta     ASR `ŋ a n a t a`           -> /akta/, with /k/
 *
 * Acronyms WITHOUT a ⟨c⟩ need no entry: `NATO` and `ASEAN` fall through the readability gate and the
 * native rules already read them correctly (nˈato, asəˈan).
 */
const ACRONYM_LEXICON: ReadonlyMap<string, string> = new Map([
    ["unesco", "unˈɛsko"],
    ["covid", "kˈovid"],
    ["acta", "ˈakta"],
]);

/** One Indonesian word → canonical IPA, RULE-ENGINE ONLY (no ⟨e⟩ lexicon). The honest, non-circular engine
 *  signal used by the referee eval; the shipped phonemizeWord layers the consensus ⟨e⟩ lexicon on top. */
export function phonemizeWordRules(word: string): string {
    // All-caps acronym → spell each letter by its Indonesian name (BBM → be-be-em → bebeem).
    //
    // ⚠ ONLY IF IT CANNOT BE READ AS A WORD, and omitting that test spelled out every acronym that is
    // actually PRONOUNCED as one: `UNESCO` came out `uɛneɛst͡ʃeo` (u-e-en-e-es-che-o) instead of
    // /unɛsko/, and `NATO`, `ASEAN`, `COVID`, `ACTA` the same way. ASEAN and UNESCO are not marginal
    // vocabulary in Indonesian or Malay. The bare `/^[A-Z]{2,}$/` claimed any capital run, so an
    // ordinary word in caps — `RUMAH` — was spelled out too.
    //
    // This is the discrimination `core/initialisms.ts` already makes for the rest of the fleet
    // (`isUnreadable`, plus a listed-exception set for the readable-but-still-spelled cases). Indonesian
    // never went through that pass, so it never got the test. Indonesian phonotactics make it clean:
    // TV / DVD / GPS / PBB have no vowel at all and are unreadable by construction, while NATO / ASEAN /
    // UNESCO syllabify without trouble.
    if (/^[A-Z]{2,}$/.test(word)) {
        const low = word.toLowerCase();
        const lex = ACRONYM_LEXICON.get(low);
        if (lex !== undefined) return lex;               // listed borrowing: its own attested reading
        if (SPELLED_ANYWAY.has(low) || isUnreadableIndonesian(low))
            return [...low].map((c) => DEF.letterNames[c] ?? "").join("");
        // readable, unlisted → fall through and read it as an ordinary word
    }
    const segs = scan(word.toLowerCase());
    if (segs.length === 0) return "";
    // Syllable-final /k/ → glottal stop [ʔ] (coda k, before a consonant or word end).
    for (let i = 0; i < segs.length; i++) {
        if (segs[i]!.ph !== "k") continue;
        const next = segs[i + 1];
        if (!next || (!VOWEL_PH.includes(next.ph[0] ?? "") && next.ph !== ""))
            segs[i]!.ph = "ʔ";
    }
    const stress = stressIndex(segs);
    let out = "";
    for (let i = 0; i < segs.length; i++) {
        if (i === stress) out += "ˈ";
        out += segs[i]!.ph;
    }
    return out.normalize("NFC");
}

// ⟨e⟩ pepet/taling lexicon — cross-source consensus (wikipron ind ∩ kaikki ind), Latin-keyed. Latin ⟨e⟩ conflates
// pepet /ə/ (the rule default) and taling /e/~/ɛ/, which is lexical and unrecoverable from the orthography; each
// entry pins the taling quality where both independent human referees agree. See indonesian-e-lexicon.tsv.
let LEXICON: Map<string, string> | undefined;
const lexicon = (): Map<string, string> =>
    (LEXICON ??= loadTsvMap(import.meta.url, "indonesian-e-lexicon.tsv", (v) => v, {
        optional: true,
    }));

/** SHIPPED Indonesian word → canonical IPA. A consensus ⟨e⟩ lexicon override resolves the pepet/taling ambiguity
 *  for known words; everything else falls through to the rule engine (which defaults pepet). */
export function phonemizeWord(word: string): string {
    if (!/^[A-Z]{2,}$/.test(word)) {
        const hit = lexicon().get(word.toLowerCase());
        if (hit !== undefined) return hit;
    }
    return phonemizeWordRules(word);
}

// ── Numbers (regular, compositional) ─────────────────────────────────────────
function numberWords(n: number): string {
    if (n < 0) return "";
    if (n < 10) return NUM.units[n]!;
    if (n < 12) return n === 10 ? NUM.seprefix + NUM.puluh : NUM.seprefix + NUM.belas;
    if (n < 20) return `${NUM.units[n - 10]} ${NUM.belas}`;
    if (n < 100) {
        const t = Math.floor(n / 10),
            r = n % 10;
        return `${NUM.units[t]} ${NUM.puluh}${r ? " " + numberWords(r) : ""}`;
    }
    if (n < 200) return `${NUM.seprefix}${NUM.ratus}${n % 100 ? " " + numberWords(n % 100) : ""}`;
    if (n < 1000) {
        const h = Math.floor(n / 100),
            r = n % 100;
        return `${NUM.units[h]} ${NUM.ratus}${r ? " " + numberWords(r) : ""}`;
    }
    if (n < 2000) return `${NUM.seprefix}${NUM.ribu}${n % 1000 ? " " + numberWords(n % 1000) : ""}`;
    if (n < 1000000) {
        const th = Math.floor(n / 1000),
            r = n % 1000;
        return `${numberWords(th)} ${NUM.ribu}${r ? " " + numberWords(r) : ""}`;
    }
    const m = Math.floor(n / 1000000),
        r = n % 1000000;
    return `${numberWords(m)} ${NUM.juta}${r ? " " + numberWords(r) : ""}`;
}

// The number class accepts Indonesian's DOT thousands grouping and COMMA decimal. Without them "9.000"
// tokenized as 9 | . | 000 and the separator became a clause PAUSE ("sembilan . nol"), and "1,5" likewise.
// Times (11.00) are claimed earlier by normalize.ts, so only real numbers reach this.
// Indonesian had no symbol tier at all: "3%" read as just "tiga", losing the percent.
const SYMBOLS = makeSymbolNormalizer({
    percent: DEF.symbolTier.percent,
    currency: DEF.symbolTier.currency,
    units: DEF.symbolTier.units,
    exponentWords: DEF.symbolTier.exponentWords,
    magnitudes: DEF.symbolTier.magnitudes,
    ampersand: DEF.symbolTier.ampersand,
    multiply: DEF.symbolTier.multiply,
});

// ⚠ THE WORD GROUP SPANS ALL OF LATIN, not just ASCII, and `[a-zA-Z]+` was silently shredding foreign names.
// A diacritic ended the token, so the letter carrying it became an unclaimed gap read as an English LETTER NAME
// and the rest of the word started over: `Cañitas` → *t͡ʃˈa ˈɛn ˈitas* ("cha EN itas"), `São` → *s ˈə ˈo*,
// `Klöcker` → *ʔl ˈoᶷ t͡ʃkˈər*. One word became three, none of them right.
// ⚠ INVISIBLE TO EVERY GATE: no digit or raw mark survives, and nothing VANISHES, so it is a WRONG-WORD defect
// that neither the leak classes nor the differential DROP test can see. Found only by reading a corpus diff.
// `\p{M}` so a DECOMPOSED accent stays with its base rather than ending the token one character later.
// yue fixed the same defect the same way and pins it (`yue("Müslüm") === phonemize(…, "en")`).
const TOKEN = /(\p{Script=Latin}[\p{Script=Latin}\p{M}]*)|([1-9]\d{0,2}(?:\.\d{3})+|\d+(?:,\d+)?)|([.?!,;:])/gu;
/** An ordinary Indonesian word is plain ASCII letters. A diacritic means a FOREIGN name — Indonesian
 *  orthography has none — so such a token goes to the injected foreign reader instead of the native g2p, which
 *  has no rule for `ñ`/`ö`/`ó` and would mangle what it cannot spell. */
const NATIVE_WORD = /^[a-zA-Z]+$/u;

class IndonesianPhonemizer implements Phonemizer {
    constructor(private foreign?: ForeignPhonemizer) {}
    text(input: string): string {
        // the Indonesian rewrites (clock, rupiah, abbreviations, slash units) run BEFORE the shared
        // symbol tier, and the clock must precede the number tokenizer so a dot-time is never read as
        // thousands grouping.
        return assembleClauses(SYMBOLS(normalizeIndonesian(input)), TOKEN, (m, sink) => {
            if (m[1]) sink.emit(
                NATIVE_WORD.test(m[1]) || this.foreign === undefined
                    ? phonemizeWord(m[1])
                    : this.foreign(m[1]),
            );
            else if (m[2]) {
                // Strip the dot grouping, then split on the decimal comma. Number words bypass the ⟨e⟩
                // lexicon (their ⟨e⟩ is pepet; avoid any taling homograph).
                const [intPart, frac] = m[2].replace(/\./gu, "").split(",");
                const n = Number(intPart);
                // ⚠ ABOVE 2^53 THIS USED TO `return` AND THE WHOLE NUMBER VANISHED FROM THE READING — the
                // decimal tail with it. Refusing to compose is right (the float has already lost the low
                // digits, so the composed numeral would be confidently wrong), but the guard had no else.
                // Digit-at-a-time is exactly what the fractional part below already does, so the fallback
                // needs no word this engine was never measured on: above 2^53 the reading is a digit
                // string, not a quantity. Inherited by `ms`/`zsm`, which wrap this engine.
                if (Number.isSafeInteger(n))
                    for (const wd of numberWords(n).split(" ")) sink.emit(phonemizeWordRules(wd));
                else
                    for (const dch of intPart!)
                        for (const wd of numberWords(Number(dch)).split(" ")) sink.emit(phonemizeWordRules(wd));
                if (frac !== undefined && frac !== "") {
                    const dot = DEF.numbers.decimalWord;
                    if (dot) sink.emit(phonemizeWordRules(dot));
                    for (const d of frac)
                        for (const wd of numberWords(Number(d)).split(" ")) sink.emit(phonemizeWordRules(wd));
                }
            } else if (m[3]) {
                const mk = CLAUSE_MARK[m[3]];
                if (mk) sink.pause(mk);
            }
        });
    }
}

/** Build the Indonesian phonemizer. */
/** `foreign` reads a token carrying a DIACRITIC — a foreign name, since Indonesian orthography has none. The
 *  registry injects English, exactly as it does for Hindi. Without it the native g2p silently drops the letter it
 *  cannot spell (`Cañitas` → *t͡ʃaˈitas*), which is quieter than the old fragmentation but no more correct. */
export function createIndonesian(foreign?: ForeignPhonemizer): Phonemizer {
    return new IndonesianPhonemizer(foreign);
}

File: csharp/Vernacula.Phonemizer/Languages/Indonesian/Indonesian.cs
/**
 * Native Indonesian (id) text phonemizer — canonical IPA.
 * Ported from src/languages/indonesian/indonesian.ts — see that file for the corpus evidence.
 */
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Indonesian;

public sealed class IndonesianNumbersDef
{
    public IReadOnlyList<string> Units { get; init; } = Array.Empty<string>();
    public string Belas { get; init; } = "";
    public string Puluh { get; init; } = "";
    public string Ratus { get; init; } = "";
    public string Ribu { get; init; } = "";
    public string Juta { get; init; } = "";
    public string Seprefix { get; init; } = "";
    /** The word for the decimal comma ("koma"). */
    public string? DecimalWord { get; init; }
}

public sealed class IndonesianDef
{
    public IReadOnlyDictionary<string, string> Digraphs { get; init; } = new Dictionary<string, string>();
    public IReadOnlyDictionary<string, string> Consonants { get; init; } = new Dictionary<string, string>();
    public IReadOnlyDictionary<string, string> Vowels { get; init; } = new Dictionary<string, string>();
    public IReadOnlyDictionary<string, string> LaxVowels { get; init; } = new Dictionary<string, string>();
    public IReadOnlyDictionary<string, string> Diphthongs { get; init; } = new Dictionary<string, string>();
    public IReadOnlyDictionary<string, string> LetterNames { get; init; } = new Dictionary<string, string>();
    public IReadOnlyDictionary<string, string> ClausePunctuation { get; init; } = new Dictionary<string, string>();
    public IndonesianNumbersDef Numbers { get; init; } = new();
    public IndonesianPhonotactics Phonotactics { get; init; } = new();
    /** The shared symbol tier's data — see the jsonc, where the evidence lives. */
    public IndonesianSymbolTier SymbolTier { get; init; } = new();
}

public sealed class IndonesianPhonotactics
{
    public string Vowels { get; init; } = "";
    public IReadOnlyList<string> Onsets { get; init; } = Array.Empty<string>();
    public IReadOnlyList<string> Codas { get; init; } = Array.Empty<string>();
    public IReadOnlyList<string> Digraphs { get; init; } = Array.Empty<string>();
}

public static class IndonesianPhonemizer
{
    public static readonly IndonesianDef DEF =
        LoadManifest.Load<IndonesianDef>("languages/indonesian", "indonesian.jsonc");
    private static IReadOnlyDictionary<string, string> CLAUSE_MARK => DEF.ClausePunctuation;
    private static IndonesianNumbersDef NUM => DEF.Numbers;

    private static bool IsVowelLetter(string c) => "aeiou".Contains(c, StringComparison.Ordinal) && c != "";
    private const string VOWEL_PH = "aeiouəɛɔɪʊ";

    private sealed class Seg
    {
        public required string Ph;
        public required string VowelLetter; // the source vowel letter (for lax resolution); "" for consonants
    }

    /** Scan a lowercased Indonesian word into segments (consonants + tense-default vowels). */
    private static List<Seg> Scan(string w)
    {
        var s = Js.CodePoints(w);
        var segs = new List<Seg>();
        string At(int k) => k >= 0 && k < s.Count ? s[k] : "";
        for (var i = 0; i < s.Count;)
        {
            var di = s[i] + At(i + 1);
            if (DEF.Diphthongs.TryGetValue(di, out var dph) && dph.Length > 0 && i + 2 == s.Count)
            {
                segs.Add(new Seg { Ph = dph, VowelLetter = "" });
                i += 2;
                continue;
            }
            var dg = s[i] + At(i + 1);
            if (DEF.Digraphs.TryGetValue(dg, out var dgh) && dgh.Length > 0)
            {
                segs.Add(new Seg { Ph = dgh, VowelLetter = "" });
                i += 2;
                continue;
            }
            var c = s[i];
            if (IsVowelLetter(c))
            {
                segs.Add(new Seg { Ph = DEF.Vowels[c], VowelLetter = c });
                i++;
            }
            else if (DEF.Consonants.TryGetValue(c, out var cph) && cph.Length > 0)
            {
                segs.Add(new Seg { Ph = cph, VowelLetter = "" });
                i++;
            }
            else i++; // unknown → skip
        }
        return segs;
    }

    private static string Head(string ph) => ph.Length > 0 ? Js.CodePoints(ph)[0] : "";

    /** Penultimate stress on the second-to-last vowel; if that nucleus is a schwa, shift to the final. */
    private static int StressIndex(List<Seg> segs)
    {
        var nuclei = segs.Select((s, i) => VOWEL_PH.Contains(Head(s.Ph), StringComparison.Ordinal) ? i : -1)
            .Where(i => i >= 0).ToList();
        if (nuclei.Count == 0) return -1;
        if (nuclei.Count == 1) return nuclei[0];
        var penult = nuclei[^2];
        if (segs[penult].Ph == "ə") return nuclei[^1]; // schwa can't bear stress
        return penult;
    }

    /** Can this letter run be an Indonesian word at all? */
    private static readonly Func<string, bool> IsUnreadableIndonesian = Initialisms.MakeUnreadableTest(new PhonotacticsData
    {
        Vowels = JsRegex.Compile($"[{DEF.Phonotactics.Vowels}]", "u"),
        LegalOnsets = new HashSet<string>(DEF.Phonotactics.Onsets, StringComparer.Ordinal),
        LegalCodas = new HashSet<string>(DEF.Phonotactics.Codas, StringComparer.Ordinal),
        Digraphs = new HashSet<string>(DEF.Phonotactics.Digraphs, StringComparer.Ordinal),
    });

    /** READABLE AND STILL SPELLED OUT — the cases the phonotactic test gets wrong in the other direction. */
    private static readonly IReadOnlySet<string> SPELLED_ANYWAY =
        new HashSet<string>(new[] { "as", "rem", "oha", "aol" }, StringComparer.Ordinal);

    /**
     * ACRONYM LEXICON — the word-pronounced borrowings whose reading is NOT derivable from Indonesian
     * orthography, and so has to be listed rather than computed.
     */
    private static readonly IReadOnlyDictionary<string, string> ACRONYM_LEXICON = new Dictionary<string, string>(StringComparer.Ordinal)
    {
        ["unesco"] = "unˈɛsko",
        ["covid"] = "kˈovid",
        ["acta"] = "ˈakta",
    };

    private static readonly JsRe ALL_CAPS = JsRegex.Compile("^[A-Z]{2,}$", "");

    /**
     * One Indonesian word → canonical IPA, RULE-ENGINE ONLY (no ⟨e⟩ lexicon). The non-circular engine signal
     * used by the referee eval; the shipped PhonemizeWord layers the consensus ⟨e⟩ lexicon on top.
     */
    public static string PhonemizeWordRules(string word)
    {
        if (ALL_CAPS.IsMatch(word))
        {
            var low = word.ToLowerInvariant();
            if (ACRONYM_LEXICON.TryGetValue(low, out var lex)) return lex; // listed borrowing: its own attested reading
            if (SPELLED_ANYWAY.Contains(low) || IsUnreadableIndonesian(low))
                return string.Concat(Js.CodePoints(low).Select(c => DEF.LetterNames.GetValueOrDefault(c) ?? ""));
        }
        var segs = Scan(word.ToLowerInvariant());
        if (segs.Count == 0) return "";
        for (var i = 0; i < segs.Count; i++)
        {
            if (segs[i].Ph != "k") continue;
            var next = i + 1 < segs.Count ? segs[i + 1] : null;
            if (next is null || (!VOWEL_PH.Contains(Head(next.Ph), StringComparison.Ordinal) && next.Ph != ""))
                segs[i].Ph = "ʔ";
        }
        var stress = StressIndex(segs);
        var @out = "";
        for (var i = 0; i < segs.Count; i++)
        {
            if (i == stress) @out += "ˈ";
            @out += segs[i].Ph;
        }
        return @out.Normalize(System.Text.NormalizationForm.FormC);
    }

    private static Dictionary<string, string>? LEXICON;
    private static Dictionary<string, string> Lexicon() =>
        LEXICON ??= LoadTsv.LoadTsvMap("languages/indonesian", "indonesian-e-lexicon.tsv", optional: true);

    /**
     * SHIPPED Indonesian word → canonical IPA. A consensus ⟨e⟩ lexicon override resolves the pepet/taling
     * ambiguity for known words; everything else falls through to the rule engine (which defaults pepet).
     */
    public static string PhonemizeWord(string word)
    {
        if (!ALL_CAPS.IsMatch(word))
        {
            if (Lexicon().TryGetValue(word.ToLowerInvariant(), out var hit)) return hit;
        }
        return PhonemizeWordRules(word);
    }

    private static string NumberWords(double n)
    {
        if (n < 0) return "";
        if (n < 10) return NUM.Units[(int)n];
        if (n < 12) return n == 10 ? NUM.Seprefix + NUM.Puluh : NUM.Seprefix + NUM.Belas;
        if (n < 20) return $"{NUM.Units[(int)n - 10]} {NUM.Belas}";
        if (n < 100)
        {
            int t = (int)(n / 10), r = (int)(n % 10);
            return $"{NUM.Units[t]} {NUM.Puluh}{(r != 0 ? " " + NumberWords(r) : "")}";
        }
        if (n < 200) return $"{NUM.Seprefix}{NUM.Ratus}{(n % 100 != 0 ? " " + NumberWords(n % 100) : "")}";
        if (n < 1000)
        {
            int h = (int)(n / 100), r = (int)(n % 100);
            return $"{NUM.Units[h]} {NUM.Ratus}{(r != 0 ? " " + NumberWords(r) : "")}";
        }
        if (n < 2000) return $"{NUM.Seprefix}{NUM.Ribu}{(n % 1000 != 0 ? " " + NumberWords(n % 1000) : "")}";
        if (n < 1000000)
        {
            double th = Math.Floor(n / 1000), r = n % 1000;
            return $"{NumberWords(th)} {NUM.Ribu}{(r != 0 ? " " + NumberWords(r) : "")}";
        }
        double m = Math.Floor(n / 1000000), rr = n % 1000000;
        return $"{NumberWords(m)} {NUM.Juta}{(rr != 0 ? " " + NumberWords(rr) : "")}";
    }

    private static readonly Func<string, string> SYMBOLS = NormalizeSymbols.MakeSymbolNormalizer(new SymbolData
    {
        Percent = DEF.SymbolTier.Percent,
        Currency = DEF.SymbolTier.Currency,
        Units = DEF.SymbolTier.Units,
        ExponentWords = DEF.SymbolTier.ExponentWords,
        Magnitudes = DEF.SymbolTier.Magnitudes,
        Ampersand = DEF.SymbolTier.Ampersand,
        Multiply = DEF.SymbolTier.Multiply,
    });

    // The word group spans ALL of Latin, not just ASCII: with `[a-zA-Z]+` a diacritic ended the token and the
    // rest of the word restarted, so one name became three. `\p{M}` keeps a DECOMPOSED accent with its base.
    // The number group accepts Indonesian's DOT thousands grouping and COMMA decimal.
    private static readonly JsRe TOKEN =
        JsRegex.Compile("(\\p{Script=Latin}[\\p{Script=Latin}\\p{M}]*)|(\\d{1,3}(?:\\.\\d{3})+|\\d+(?:,\\d+)?)|([.?!,;:])", "gu");
    /**
     * An ordinary Indonesian word is plain ASCII letters. A diacritic means a FOREIGN name — Indonesian
     * orthography has none — so such a token goes to the injected foreign reader instead of the native g2p,
     * which has no rule for `ñ`/`ö`/`ó` and would mangle what it cannot spell.
     */
    private static readonly JsRe NATIVE_WORD = JsRegex.Compile("^[a-zA-Z]+$", "u");

    private static readonly JsRe GROUPING_DOT = JsRegex.Compile("\\.", "gu");

    private sealed class Engine : ILanguage
    {
        private readonly Func<string, string>? _foreign;
        internal Engine(Func<string, string>? foreign = null) => _foreign = foreign;

        public string Text(string input)
        {
            // Order: the Indonesian rewrites (clock, rupiah, abbreviations, slash units) run BEFORE the shared
            // symbol tier, and the clock must precede the number tokenizer so a dot-time is never read as
            // thousands grouping.
            return Clauses.AssembleClauses(SYMBOLS(Normalize.NormalizeIndonesian(input)), TOKEN, (m, sink) =>
            {
                if (m.Groups[1].Success && m.Groups[1].Value.Length > 0)
                    sink.Emit(NATIVE_WORD.IsMatch(m.Groups[1].Value) || _foreign is null
                        ? PhonemizeWord(m.Groups[1].Value)
                        : _foreign(m.Groups[1].Value));
                else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0)
                {
                    // Strip the dot grouping, then split on the decimal comma. Number words deliberately
                    // bypass the ⟨e⟩ lexicon (PhonemizeWordRules): their ⟨e⟩ is pepet, and a taling homograph
                    // would win otherwise.
                    var split = GROUPING_DOT.Replace(m.Groups[2].Value, "").Split(',');
                    var intPart = split[0];
                    var frac = split.Length > 1 ? split[1] : null;
                    var n = Js.Number(intPart);
                    // `n` is a JS `number` here as in the TS: past 2^53 the low digits are already gone, so
                    // composing a numeral would be confidently wrong. Digit-at-a-time instead — never a bare
                    // `return`, which used to make the whole number vanish from the reading.
                    if (double.IsInteger(n) && Math.Abs(n) <= 9007199254740991d)
                        foreach (var wd in NumberWords(n).Split(' ')) sink.Emit(PhonemizeWordRules(wd));
                    else
                        foreach (var dch in intPart)
                            foreach (var wd in NumberWords(Js.Number(dch.ToString())).Split(' ')) sink.Emit(PhonemizeWordRules(wd));
                    if (frac is not null && frac != "")
                    {
                        var dot = DEF.Numbers.DecimalWord;
                        if (!string.IsNullOrEmpty(dot)) sink.Emit(PhonemizeWordRules(dot));
                        foreach (var d in frac)
                            foreach (var wd in NumberWords(Js.Number(d.ToString())).Split(' ')) sink.Emit(PhonemizeWordRules(wd));
                    }
                }
                else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0)
                {
                    var mk = CLAUSE_MARK.GetValueOrDefault(m.Groups[3].Value);
                    if (mk is not null) sink.Pause(mk);
                }
            });
        }
    }

    /**
     * Build the Indonesian phonemizer. `foreign` reads a token carrying a DIACRITIC — a foreign name, since
     * Indonesian orthography has none; the registry injects English. Without it the native g2p silently drops
     * the letter it cannot spell.
     */
    public static ILanguage CreateIndonesian(Func<string, string>? foreign = null) => new Engine(foreign);

    internal static void RegisterSelf() =>
        Registry.Register("indonesian", () => CreateIndonesian(Registry.ReadAsEnglish));
}

public sealed class IndonesianSymbolTier
{
    public IReadOnlyList<string> Percent { get; init; } = Array.Empty<string>();
    public IReadOnlyDictionary<string, IReadOnlyList<string>> Currency { get; init; } = new Dictionary<string, IReadOnlyList<string>>();
    public IReadOnlyDictionary<string, IReadOnlyList<string>> Units { get; init; } = new Dictionary<string, IReadOnlyList<string>>();
    public ExponentWordsDef ExponentWords { get; init; } = new();
    public IReadOnlyList<string> Magnitudes { get; init; } = Array.Empty<string>();
    public string Ampersand { get; init; } = "";
    public MultiplyDef Multiply { get; init; } = null!;
}
